File: ycmd/completers/all/tests/identifier_utils_test.py
#!/usr/bin/env python
#
# Copyright (C) 2013  Google Inc.
#
# This file is part of YouCompleteMe.
#
# YouCompleteMe is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# YouCompleteMe is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with YouCompleteMe.  If not, see <http://www.gnu.org/licenses/>.

from nose.tools import eq_
from ycmd.completers.all import identifier_utils


def RemoveIdentifierFreeText_CppComments_test():
  eq_( "foo \nbar \nqux",
       identifier_utils.RemoveIdentifierFreeText(
          "foo \nbar //foo \nqux" ) )


def RemoveIdentifierFreeText_PythonComments_test():
  eq_( "foo \nbar \nqux",
       identifier_utils.RemoveIdentifierFreeText(
          "foo \nbar #foo \nqux" ) )


def RemoveIdentifierFreeText_CstyleComments_test():
  eq_( "foo \nbar \nqux",
       identifier_utils.RemoveIdentifierFreeText(
          "foo \nbar /* foo */\nqux" ) )

  eq_( "foo \nbar \nqux",
       identifier_utils.RemoveIdentifierFreeText(
          "foo \nbar /* foo \n foo2 */\nqux" ) )


def RemoveIdentifierFreeText_SimpleSingleQuoteString_test():
  eq_( "foo \nbar \nqux",
       identifier_utils.RemoveIdentifierFreeText(
          "foo \nbar 'foo'\nqux" ) )


def RemoveIdentifierFreeText_SimpleDoubleQuoteString_test():
  eq_( "foo \nbar \nqux",
       identifier_utils.RemoveIdentifierFreeText(
          'foo \nbar "foo"\nqux' ) )


def RemoveIdentifierFreeText_EscapedQuotes_test():
  eq_( "foo \nbar \nqux",
       identifier_utils.RemoveIdentifierFreeText(
          "foo \nbar 'fo\\'oz\\nfoo'\nqux" ) )

  eq_( "foo \nbar \nqux",
       identifier_utils.RemoveIdentifierFreeText(
          'foo \nbar "fo\\"oz\\nfoo"\nqux' ) )


def RemoveIdentifierFreeText_SlashesInStrings_test():
  eq_( "foo \nbar baz\nqux ",
       identifier_utils.RemoveIdentifierFreeText(
           'foo \nbar "fo\\\\"baz\nqux "qwe"' ) )

  eq_( "foo \nbar \nqux ",
       identifier_utils.RemoveIdentifierFreeText(
           "foo '\\\\'\nbar '\\\\'\nqux '\\\\'" ) )


def RemoveIdentifierFreeText_EscapedQuotesStartStrings_test():
  eq_( "\\\"foo\\\" zoo",
       identifier_utils.RemoveIdentifierFreeText(
           "\\\"foo\\\"'\"''bar' zoo'test'" ) )

  eq_( "\\'foo\\' zoo",
       identifier_utils.RemoveIdentifierFreeText(
           "\\'foo\\'\"'\"\"bar\" zoo\"test\"" ) )


def RemoveIdentifierFreeText_NoMultilineString_test():
  eq_( "'\nlet x = \nlet y = ",
       identifier_utils.RemoveIdentifierFreeText(
           "'\nlet x = 'foo'\nlet y = 'bar'" ) )

  eq_( "\"\nlet x = \nlet y = ",
       identifier_utils.RemoveIdentifierFreeText(
           "\"\nlet x = \"foo\"\nlet y = \"bar\"" ) )


def RemoveIdentifierFreeText_PythonMultilineString_test():
  eq_( "\nzoo",
       identifier_utils.RemoveIdentifierFreeText(
           "\"\"\"\nfoobar\n\"\"\"\nzoo" ) )

  eq_( "\nzoo",
       identifier_utils.RemoveIdentifierFreeText(
           "'''\nfoobar\n'''\nzoo" ) )


def ExtractIdentifiersFromText_test():
  eq_( [ "foo", "_bar", "BazGoo", "FOO", "_", "x", "one", "two", "moo", "qqq" ],
       identifier_utils.ExtractIdentifiersFromText(
           "foo $_bar \n&BazGoo\n FOO= !!! '-' - _ (x) one-two !moo [qqq]" ) )

File: cpp/ycm/tests/IdentifierUtils_test.cpp
// Copyright (C) 2011, 2012 Google Inc.
//
// This file is part of ycmd.
//
// ycmd is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// ycmd is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with ycmd.  If not, see <http://www.gnu.org/licenses/>.

#include "IdentifierUtils.h"
#include "TestUtils.h"
#include "IdentifierDatabase.h"

#include <gtest/gtest.h>
#include <gmock/gmock.h>
#include <boost/filesystem.hpp>

namespace YouCompleteMe {

namespace fs = boost::filesystem;
using ::testing::ElementsAre;
using ::testing::ContainerEq;
using ::testing::WhenSorted;
using ::testing::Pair;
using ::testing::UnorderedElementsAre;


TEST( IdentifierUtilsTest, ExtractIdentifiersFromTagsFileWorks ) {
  fs::path root = fs::current_path().root_path();
  fs::path testfile = PathToTestFile( "basic.tags" );
  fs::path testfile_parent = testfile.parent_path();

  EXPECT_THAT( ExtractIdentifiersFromTagsFile( testfile ),
      UnorderedElementsAre(
        Pair( "cpp", UnorderedElementsAre(
                         Pair( ( testfile_parent / "foo" ).string(),
                               ElementsAre( "i1", "foosy" ) ),
                         Pair( ( testfile_parent / "bar" ).string(),
                               ElementsAre( "i1", "fooaaa" ) ) ) ),
        Pair( "fakelang", UnorderedElementsAre(
                              Pair( ( root / "foo" ).string(),
                                    ElementsAre( "zoro" ) ) ) ),
        Pair( "cs", UnorderedElementsAre(
                        Pair( ( root / "m_oo" ).string(),
                              ElementsAre( "#bleh" ) ) ) ),
        Pair( "foobar", UnorderedElementsAre(
                            Pair( ( testfile_parent / "foo.bar" ).string(),
                                  ElementsAre( "API", "DELETE" ) ) ) ),
        Pair( "c", UnorderedElementsAre(
                       Pair( ( root / "foo" / "zoo" ).string(),
                             ElementsAre( "Floo::goo" ) ),
                       Pair( ( root / "foo" / "goo maa" ).string(),
                             ElementsAre( "!goo" ) ) ) ) ) );
}


TEST( IdentifierUtilsTest, TagFileIsDirectory ) {
  fs::path testfile = PathToTestFile( "directory.tags" );

  EXPECT_THAT( ExtractIdentifiersFromTagsFile( testfile ),
               ContainerEq( FiletypeIdentifierMap() ) );
}


TEST( IdentifierUtilsTest, TagFileIsEmpty ) {
  fs::path testfile = PathToTestFile( "empty.tags" );

  EXPECT_THAT( ExtractIdentifiersFromTagsFile( testfile ),
               ContainerEq( FiletypeIdentifierMap() ) );
}


TEST( IdentifierUtilsTest, TagLanguageMissing ) {
  fs::path testfile = PathToTestFile( "invalid_tag_file_format.tags" );

  EXPECT_THAT( ExtractIdentifiersFromTagsFile( testfile ),
               ContainerEq( FiletypeIdentifierMap() ) );
}


TEST( IdentifierUtilsTest, TagFileInvalidPath ) {
  fs::path testfile = PathToTestFile( "invalid_path_to_tag_file.tags" );

  EXPECT_THAT( ExtractIdentifiersFromTagsFile( testfile ),
               ContainerEq( FiletypeIdentifierMap() ) );
}

} // namespace YouCompleteMe

